dali/pipeline/operator/false_gpu_operator.h

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATOR_FALSE_GPU_OPERATOR_H_
+#define DALI_PIPELINE_OPERATOR_FALSE_GPU_OPERATOR_H_
+
+#include <limits>
+#include <memory>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+#include "dali/pipeline/data/tensor.h"
+#include "dali/pipeline/data/views.h"
+#include "dali/pipeline/operator/argument.h"
+#include "dali/pipeline/operator/common.h"
+#include "dali/pipeline/operator/op_spec.h"
+#include "dali/pipeline/operator/operator.h"
+#include "dali/pipeline/operator/sequence_shape.h"
+
+namespace dali {
+
+/**
+ * @brief Produces a false GPU version of a given CPU operator.
+ *        The resulting operator copies GPU inputs to host memory before setup, and copies
+ *        the CPU outputs to device memory after Run.
+ *
+ * @tparam CPUOperator CPU operator class
+ */
+template <typename CPUOperator>
+class FalseGPUOperator : public Operator<GPUBackend> {
+ public:
+  explicit FalseGPUOperator(const OpSpec &spec)
+      : Operator<GPUBackend>(spec),
+        cpu_impl_(spec),
+        thread_pool_(num_threads_, spec.GetArgument<int>("device_id"), true /** set_affine */ ) {
+    cpu_ws_.SetThreadPool(&thread_pool_);
+  }
+  ~FalseGPUOperator() override = default;
+
+ protected:
+  bool CanInferOutputs() const override {
+    // To run Setup we need to first copy from device to host.
```

Comment:
Do we need to copy all data or just CPU arguments?
I have mixed feelings regarding moving the whole setup to Run, but it is up to you.

Reply:
To run Setup we need to have a valid HostWorkspace